dataprep/clean/clean_df.py

```diff
@@ -0,0 +1,414 @@
+    )
+    column_not_na_subset = column_not_na.sample(n=sample_size, random_state=1)
+
+    # 1. For geographic coordinates: lat and long
+    lat_lon = ["lat", "Lat", "LAT", "lon", "Lon", "LON"]
+    if any(x in column.name for x in lat_lon):
+        lat_long_infer_count = sum(
+            pd.Series(validate_lat_long(column_not_na_subset, lat_long=False, lat=True)).tolist()
+        )
+        # return if at least half of the entries match
+        if lat_long_infer_count > (column_not_na_subset.size / 2):
+            return "coordinate"
+
+    # 2. For numeric ("integer", "floating") or boolean data types
+    if infer_dtype(column_not_na_subset) != "string":
+        return infer_dtype(column_not_na_subset)
+
+    # 3. For string and semantic data types (email, country, phone, etc.)
+    default_infer_dtype = infer_dtype(column_not_na_subset)
+
+    semantic_data_type_dic = {"email": 0, "country": 0, "phone": 0, "ip": 0, "URL": 0, "address": 0}
+
+    semantic_data_type_dic["email"] = sum(pd.Series(validate_email(column_not_na_subset)).tolist())
+    semantic_data_type_dic["country"] = sum(
+        pd.Series(validate_country(column_not_na_subset)).tolist()
+    )
+    semantic_data_type_dic["phone"] = sum(pd.Series(validate_phone(column_not_na_subset)).tolist())
+    semantic_data_type_dic["ip"] = sum(pd.Series(validate_ip(column_not_na_subset)).tolist())
+    semantic_data_type_dic["URL"] = sum(pd.Series(validate_url(column_not_na_subset)).tolist())
+    semantic_data_type_dic["address"] = sum(
+        pd.Series(validate_address(column_not_na_subset)).tolist()
+    )
+
+    if all(value == 0 for value in semantic_data_type_dic.values()):
+        # no semantic data types match, return the default one
+        return default_infer_dtype
+    # else, return the best inferred semantic data type
+    best_infer = max(semantic_data_type_dic.items(), key=operator.itemgetter(1))[0]
+    best_infer_count = max(semantic_data_type_dic.items(), key=operator.itemgetter(1))[1]
+    return best_infer if best_infer_count > (column_not_na_subset.size / 2) else default_infer_dtype
+
+
+def _infer_atomic_data_type(column: pd.Series) -> Any:
+    """
+    Function to infer the atomic data type for a column.
+
+    Parameters
+    ----------
+    column
+        A Column of a Pandas DataFrame to be checked.
+    """
+
+    return infer_dtype(column[column.apply(_check_valid_values, 0)])
+
+
+def _infer_data_type_df(df: pd.DataFrame, data_type_detection: str) -> pd.DataFrame:
+    """
+    Function to infer the best data type for a whole data frame.
+
+    Parameters
+    ----------
+    df
+        A Pandas dataframe containing the data.
+    data_type_detection
+        The desired way to check data types.
+        * If ’semantic’, then perform a column-wise semantic and atomic type detection.
+        * If 'atomic', then return the best inferred atomic data type from Python default ones.
+        * If 'none', then no results will be returned.
```

Comment:
I think the input can't be 'none' here

Reply:
I guess "none" means that the users do not want to perform any data type detection?

Reply:
> I guess "none" means that the users do not want to perform any data type detection?

Based on my understanding of the implementation here, the input parameter of `_infer_data_type_df()` will not be 'none'. If it's not 'semantic', you will perform the atomic detection. Which means it can only be 'semantic' and 'atomic'?

Reply:
> > I guess "none" means that the users do not want to perform any data type detection?
> 
> Based on my understanding of the implementation here, the input parameter of `_infer_data_type_df()` will not be 'none'. If it's not 'semantic', you will perform the atomic detection. Which means it can only be 'semantic' and 'atomic'?

Yeah, I agree with Yi's opinion. It seems that the `_infer_data_type_df()` function doesn't have an option for the parameter 'none'. 
<img width="798" alt="Screen Shot 2021-04-20 at 4 32 15 PM" src="https://user-images.githubusercontent.com/10005367/115364433-f9816b80-a1f5-11eb-9c23-a3f71bd12b11.png">
How about changing the code like this?
<img width="797" alt="Screen Shot 2021-04-20 at 4 36 17 PM" src="https://user-images.githubusercontent.com/10005367/115364997-8a584700-a1f6-11eb-95f5-350a7abdf82e.png">

Reply:
Thanks Danrui! Greatly appreciated. I did not find this typo before.